doc/app_devel_docs.rst

```diff
@@ -0,0 +1,337 @@
+    ev <- switchPromptlyDyn dynEv
+
+  The biggest problem with such errors is that the line numbers are not correct, so it can take a while to figure out the source of error
+
+  One possible solution is to explicitly specify the type of functions and expression in the ``let`` and ``do`` block inside of ``rec``::
+
+    --  This is required to specify the types
+    --  {-# LANGUAGE ScopedTypeVariables #-}
+
+    -- This can be useful to specify types partially, just to help figure out source of error
+    --  {-# LANGUAGE PartialTypeSignatures #-}
+
+    -- Specify an explicit forall
+    myWidget :: forall t m k . (MonadWidget t m, Ord k)
+      => Map k Text -> m ()
+    myWidget mapInput = do
+      ..
+
+      rec
+        let
+          eTabClicks :: Event t k = leftmost tabClicksList
+
+        d :: Dynamic t k <- do
+          someCodeThatIsSupposedToReturnDynamicK
+
+.. _ffi:
+
+Web APIs and FFI
+----------------
+
+* For working with DOM and using Web APIs the ``ghcjs-dom`` package should suffice.
+
+  It provides APIs like ``getElementById``, ``getBoundingRect`` to work with DOM, and many other Web APIs related to geolocation, media management, web audio, etc.
+
+  To use the DOM related APIs for ``reflex-dom`` created elements, extract the `raw` element from the `reflex element` ::
+
+    import qualified GHCJS.DOM.Types as DOM
+    import qualified GHCJS.DOM.DOMRectReadOnly as DOM
+    import qualified GHCJS.DOM.Element as DOM
+
+      (_,e) <- el' "div" $ text "Hello"
+      
+      let getCoords e = DOM.liftJSM $ do
+            rect <- DOM.getBoundingClientRect (_raw_element e)
+            y <- DOM.getY rect
+            h <- DOM.getHeight rect
+            return (y,h)
+
+      performEvent (getCoords e <$ ev)
+
+* But when using external .js files, one has to do arbitrary JS code execution.
+
+  For doing this ``jsaddle`` package is preferred as it provides a type-safe way to execute the JS code.
+
+  See `documentation <https://hackage.haskell.org/package/jsaddle-0.9.4.0/docs/Language-Javascript-JSaddle-Object.html>`_ of ``Language.Javascript.JSaddle.Object`` for examples
+
+  See :ref:`dom_ui_libs` for example usage.
+
+* It is also possible to do arbitrary JS code block execution using ``eval`` API from ``Language.Javascript.JSaddle.Evaluate``. ::
+
+    eval :: (ToJSString script) => script -> JSM JSVal
+
+    liftJSM $ eval "console.log('Hello World')"
+
+* JSFFI functions
+
+  This will only work with ``ghcjs``::
+
+    import GHCJS.Types (JSVal)
+    
+    foreign import javascript unsafe
+      "try { $r = $1 / $2; } catch (e) { $r = "error"; }"
+      divide :: Double -> Double -> JSVal
+
+  See https://github.com/ghcjs/ghcjs/blob/master/doc/foreign-function-interface.md
+
+If you access a DOM element via FFI which has just been created, then it might result in an error. This is because the ``getPostBuild`` is fired before the DOM has been put in the Window. To solve this add ``delay``::
```

Comment:
`delay` should be avoided as well. @ryantrinkle isn't there a better solution for this?

Reply:
I'd love this too. Perhaps a `getPostRender` event? Could aid my experiments in this direction greatly: https://github.com/Wizek/reflex-dom-testing/blob/94f3380/frontend/src/Main.hs#L70-L125

Reply:
@Wizek [This](https://github.com/reflex-frp/reflex-dom/pull/152) introduces mount state reporting but it has a performance penalty. More work must be done to get rid of the slowdown but the kernel is there.

Reply:
@ryantrinkle do you have a better solution/alternative to this?

Reply:
Removed this part.